_drafts/2023-08-15-debugging-android-devices.md

```diff
@@ -0,0 +1,274 @@
+
+ Please refer to [Android's official documentation](https://developer.android.com/topic/performance/vitals/anr) for more details.
+
+### Java Exceptions
+
+This is the most common type of crash that most Android developers are accustomed to in Android apps and services. In Java, and like many other languages, an exception is an event that occurs during the execution of a program, disrupting the normal flow of instructions. It indicates that something unexpected or erroneous has happened, preventing the program from continuing its normal execution. If you're looking for more information about reading and understanding Java-based exceptions, I highly recommend reading [Twilio's excellent blog post](https://www.twilio.com/blog/how-to-read-and-understand-a-java-stacktrace) that goes into great depth about them.
+
+### WTFs
+
+[What Terrible Failures](https://developer.android.com/reference/android/util/Log.html#wtf(java.lang.String,%20java.lang.String)) according to Google should be used to:
+
+> Report a condition that should never happen. The error will always be logged at level ASSERT with the call stack. Depending on system configuration, a report may be added to the DropBoxManager, and/or the process may be terminated immediately with an error dialog.
+
+These are logged in Android apps and services throughout the system to record things that as the documentation says, should never happen but we all know that things that should never happen on our systems do still happen, and that's the perfect time to trigger a WTF.
+
+### SeLinux Violations
+
+Android's SELinux (Security-Enhanced Linux) is an implementation of the SELinux framework specifically created by the National Security Agency (NSA). SELinux is used on Android devices to enhance security by enforcing mandatory access controls and ensuring that processes and applications operate within defined security contexts. Android's implementation of SELinux is designed to provide a robust security mechanism to mitigate risks associated with malware, unauthorized access, and other security threats. When incorrectly configured, violations will be logged to logcat via `auditd`. Violations can lead to crashes when the resources attempting to be accessed are protected. One important detail to remember is that in engineering builds, SSLinux is configured to be permissive and will only log a warning but will still allow access to the resource being requested. This can make it easy to miss violations during development.
+
+You can find more details here in the [official documentation](https://source.android.com/docs/security/features/selinux/).
+
+## Diagnostic Tools
+
+Due to the complexity of Android-based devices, Google ships several tools to help diagnose what is happening on them.
+
+### Android Debug Bridge (ADB)
```

Comment:
This feels a bit weird being in the middle of the other things - it's the main way we get at everything else mentioned; should it be either at the top ("here's how you connect to a device - then what you can do once connected") or at the bottom ("here's how you can get at all the stuff we just went through").

Reply:
yeah true, I'll reword it some to emphasize its importance and move it right above the conclusion.